### PR TITLE
[omnibus] Farewell APT

### PR DESCRIPTION
**Pull request by 0intro**
### What does this PR do?

This change rewrites the dpkginfo probe in OpenSCAP so it doesn't depend on the APT library anymore.

The dpkginfo probe now parses the `/var/lib/dpkg/status` file using a memory efficient method. This file contains the list of installed packages.

Consequently, the APT library has been removed from omnibus.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: omnibus/config/patches/openscap/dpkginfo-no-apt.patch [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
++			if (strncmp(key, "Status", 6) == 0) {
++				if (strncmp(value, "install", 7) != 0) {
++					// Package deinstalled.
++					dpkginfo_free_reply(reply);
++					reply = NULL;
++					continue;
++				}
++			}
++			if (strncmp(key, "Architecture", 12) == 0)
++				reply->arch = strdup(value);
```
Comment (chouquette) [inline omnibus/config/patches/openscap/dpkginfo-no-apt.patch:234]: nit: you could `continue` here

Comment (chouquette) [inline omnibus/config/patches/openscap/dpkginfo-no-apt.patch:234]: nit²: no allocation check

File: omnibus/config/patches/openscap/dpkginfo-no-apt.patch [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+index e07aa2fe4..72f593425 100644
+--- a/src/OVAL/probes/unix/linux/dpkginfo-helper.h
++++ b/src/OVAL/probes/unix/linux/dpkginfo-helper.h
+@@ -22,10 +22,6 @@
+ #ifndef __DPKGINFO_HELPER__
+ #define __DPKGINFO_HELPER__
+ 
+-#ifdef __cplusplus
+-extern "C" {
+-#endif
```
Comment (chouquette) [inline omnibus/config/patches/openscap/dpkginfo-no-apt.patch:275]: Isn't this still needed?

File: omnibus/config/software/apt.rb [hunk 3]
```diff
@@ -1,99 +0,0 @@
...
-default_version '2.7.3'
-
-dependency 'bzip2'
-dependency 'gnutls'
-dependency 'libdb'
-dependency 'libgcrypt'
-dependency 'libiconv'
-dependency 'liblz4'
-dependency 'liblzma'
-dependency 'libxxhash'
```
Comment (chouquette) [inline omnibus/config/software/apt.rb:18]: AFAICS this dependency isn't needed anymore

File: omnibus/config/software/apt.rb [hunk 4]
```diff
@@ -1,99 +0,0 @@
...
-
-name 'apt'
-default_version '2.7.3'
-
-dependency 'bzip2'
-dependency 'gnutls'
-dependency 'libdb'
-dependency 'libgcrypt'
-dependency 'libiconv'
-dependency 'liblz4'
```
Comment (chouquette) [inline omnibus/config/software/apt.rb:16]: This was only used by APT, AFAICS

**Review by chouquette — COMMENTED** [review]
Qutie a few nitpicks/general questions but I really like the idea or removing our dependency to libapt :)
I'll let you decide if you prefer to remove libapt only dependency in this PR or as a follow up, as long as they are also removed

File: omnibus/config/software/apt.rb (hunk 3 above)
Comment (0intro) [reply omnibus/config/software/apt.rb:18]: Done.

File: omnibus/config/software/apt.rb (hunk 4 above)
Comment (0intro) [reply omnibus/config/software/apt.rb:16]: Done.

**Review by chouquette — CHANGES_REQUESTED** [review]

File: omnibus/config/patches/openscap/dpkginfo-no-apt.patch [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
++		if (buf[0] == '\n') {
++			// New package entry.
++			if (reply != NULL) {
++				// Package found.
++				*err = 1;
++				goto out;
++			}
++			continue;
++		}
++		if (buf[0] == ' ') {
```
[the remark this answers is not in the document]
Comment (0intro) [reply omnibus/config/patches/openscap/dpkginfo-no-apt.patch:216]: Done.

File: omnibus/config/patches/openscap/dpkginfo-no-apt.patch [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
++	if (reply != NULL) {
++		// Package found.
++		*err = 1;
++	}
++
++out:
++	fclose(f);
++	return reply;
++err:
++	fclose(f);
```
[the remark this answers is not in the document]
Comment (0intro) [reply omnibus/config/patches/openscap/dpkginfo-no-apt.patch:273]: Done.

**Review by chouquette — APPROVED** [review]
LGTM, thanks!

**Review by paulcacheux — APPROVED** [review]

**Pull request merged**